hudi-spark-datasource/hudi-spark/src/main/scala/org/apache/hudi/IncrementalRelation.scala

```diff
@@ -155,27 +155,60 @@ class IncrementalRelation(val sqlContext: SQLContext,
     if (filteredRegularFullPaths.isEmpty && filteredMetaBootstrapFullPaths.isEmpty) {
       sqlContext.sparkContext.emptyRDD[Row]
     } else {
-      log.info("Additional Filters to be applied to incremental source are :" + filters)
+      log.info("Additional Filters to be applied to incremental source are :" + filters.mkString("Array(", ", ", ")"))
 
       var df: DataFrame = sqlContext.createDataFrame(sqlContext.sparkContext.emptyRDD[Row], usedSchema)
 
-      if (metaBootstrapFileIdToFullPath.nonEmpty) {
-        df = sqlContext.sparkSession.read
-               .format("hudi")
-               .schema(usedSchema)
-               .option(DataSourceReadOptions.READ_PATHS.key, filteredMetaBootstrapFullPaths.mkString(","))
-               .load()
+      val fullTableScanFallback = optParams.getOrElse(DataSourceReadOptions.INCREMENTAL_FALLBACK_TO_FULL_TABLE_SCAN_FOR_NON_EXISTING_FILES.key,
+        DataSourceReadOptions.INCREMENTAL_FALLBACK_TO_FULL_TABLE_SCAN_FOR_NON_EXISTING_FILES.defaultValue).toBoolean
+
+      var doFullTableScan = false
+
+      if (fullTableScanFallback) {
+        val fs = new Path(basePath).getFileSystem(sqlContext.sparkContext.hadoopConfiguration);
+        val timer = new HoodieTimer().startTimer();
+
+        val allFilesToCheck = filteredMetaBootstrapFullPaths ++ filteredRegularFullPaths
+        val firstNotFoundPath = allFilesToCheck.find(path => !fs.exists(new Path(path)))
+        val timeTaken = timer.endTimer()
+        log.info("Checking if paths exists took " + timeTaken + "ms")
+
+        val optStartTs = optParams(DataSourceReadOptions.BEGIN_INSTANTTIME.key)
+        val isInstantArchived = optStartTs.compareTo(commitTimeline.firstInstant().get().getTimestamp) < 0 // True if optStartTs < activeTimeline.first
```

Comment:
From a user standpoint, I would expect we should fallback to first valid commit in active timeline which cleaner has not cleaned up. But guess from an impl standpoint, we can't find this commit that easily. And so is the rational to fallback to snapshot query?

Reply:
my bad. from re-reading the description, guess the fix does not sit well. cleaner will not touch the timeline right. So, how do we know if a commit has been cleaned up or not (bcoz, it could still be part of active timeline)

Reply:
I revisited this patch. I get it now. 
So, we are fixing two things. 
1: a commit is valid in active timeline, but corresponding data files are cleaned up.
2: begin commit is archived. 
Makes sense to me.